numpy/_core/src/umath/ufunc_object.c

```diff
@@ -2829,18 +2837,17 @@ PyUFunc_Accumulate(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
              * Output (dataptr[0]) and input (dataptr[1]) may point to
              * the same memory, e.g. np.add.accumulate(a, out=a).
              */
-            if (descrs[2]->type_num == NPY_OBJECT) {
-                /*
-                 * Incref before decref to avoid the possibility of the
-                 * reference count being zero temporarily.
-                 */
-                Py_XINCREF(*(PyObject **)dataptr_copy[1]);
-                Py_XDECREF(*(PyObject **)dataptr_copy[0]);
-                *(PyObject **)dataptr_copy[0] =
-                                    *(PyObject **)dataptr_copy[1];
+            if (copy_info.func) {
+                const npy_intp one = 1;
+                if (copy_info.func(
+                        &copy_info.context, &dataptr_copy[1], &one,
+                        &stride_copy[1], copy_info.auxdata) < 0) {
+                    NPY_END_THREADS;
+                    goto fail;
+                }
             }
             else {
-                memmove(dataptr_copy[0], dataptr_copy[1], itemsize);
+                memmove(dataptr_copy[2], dataptr_copy[1], itemsize);
```

Comment:
why this change for the non-cast case?

Reply:
Just because for the other path `dataptr_copy` entries 1 and 2 are more natural (due to ordering), so I thought it is clearer to use also 1 and 2 here.